src/Corax/Mappings/IndexFieldsMappingBuilder.cs

```diff
@@ -0,0 +1,185 @@
+        _fields = new(SliceComparer.Instance);
+        _fieldsById = new();
+        _isForWriter = isForWriter;
+        _isDynamic = isDynamic;
+    }
+
+    public IndexFieldsMappingBuilder AddBindingToEnd(string fieldName, Analyzer analyzer = null, bool hasSuggestion = false,
+        FieldIndexingMode fieldIndexingMode = FieldIndexingMode.Normal, bool hasSpatial = false)
+    {
+        Slice.From(_context, fieldName, out var slice);
+        return AddBinding(_fields.Count, slice, analyzer, hasSuggestion, fieldIndexingMode, hasSpatial);
+    }
+
+    public IndexFieldsMappingBuilder AddBindingToEnd(Slice fieldName, Analyzer analyzer = null, bool hasSuggestion = false,
+        FieldIndexingMode fieldIndexingMode = FieldIndexingMode.Normal, bool hasSpatial = false) =>
+        AddBinding(_fields.Count, fieldName, analyzer, hasSuggestion, fieldIndexingMode, hasSpatial);
+
+    public IndexFieldsMappingBuilder AddDynamicBinding(Slice fieldName, FieldIndexingMode mode)
+    {
+        if (_fields.TryGetValue(fieldName, out var storedBinding) == false)
+        {
+            AddBindingToEnd(fieldName, GetAnalyzer(), fieldIndexingMode: mode);
+        }
+        else
+        {
+            Debug.Assert(storedBinding.FieldIndexingMode == mode);
+        }
+
+        Analyzer GetAnalyzer() => mode switch
+        {
+            FieldIndexingMode.Search => _searchAnalyzer(fieldName.ToString()),
+            FieldIndexingMode.Exact => _exactAnalyzer(fieldName.ToString()),
+            FieldIndexingMode.No => null,
+            _ => _defaultAnalyzer
+        };
+
+        return this;
+    }
+    
+    public IndexFieldsMappingBuilder AddBinding(int fieldId, string fieldName, Analyzer analyzer = null, bool hasSuggestion = false,
+        FieldIndexingMode fieldIndexingMode = FieldIndexingMode.Normal, bool hasSpatial = false)
+    {
+        Slice.From(_context, fieldName, out var slice);
+        return AddBinding(fieldId, slice, analyzer, hasSuggestion, fieldIndexingMode, hasSpatial);
+    }
+    
+    public IndexFieldsMappingBuilder AddBinding(int fieldId, Slice fieldName, Analyzer analyzer = null, bool hasSuggestion = false,
+        FieldIndexingMode fieldIndexingMode = FieldIndexingMode.Normal, bool hasSpatial = false)
+    {
+        if (_fieldsById.TryGetValue(fieldId, out var storedAnalyzer) == false)
+        {
+            var clonedFieldName = fieldName.Clone(_context);
+            GetFieldNameForDoubles(clonedFieldName, out var fieldNameDouble);
+            GetFieldNameForLongs(clonedFieldName, out var fieldNameLong);
+            var binding = new IndexFieldBinding(fieldId, clonedFieldName, fieldNameLong, fieldNameDouble, _isForWriter,
+                analyzer, hasSuggestion, fieldIndexingMode, hasSpatial);
+            _fields[clonedFieldName] = binding;
+            _fieldsById[fieldId] = binding;
+        }
+        else
+        {
+            Debug.Assert(analyzer == storedAnalyzer.Analyzer);
```

Comment:
Please add message (2nd parameter) telling us what were the analyzers so we'll know that once if fails